### PR TITLE
make height of remaining libs modules configurable/refactor

### PR DESCRIPTION
**Pull request by dterrahe**
Allow setting a maximum on the widget height (using ctrl-scroll) of all the remaining lib modules in lighttable and darkroom that didn't have this feature yet;
history
duplicates
colorpicker
snapshots
styles
recent collections
and refactored metadata and masks (which had coincidentally just a few days back received this feature from @johnny-bit)

The functionality here is slightly different from elsewhere in that there is no hard minimum size; if you want to, you can force the list to be one line high. However if you allow it to be higher than the hardcoded "minimum" size, it will only use the space required for the current number of entries. Also, the height, and step size, when using the mouse wheel to scroll on the list itself, not the scrollbar, will always be a multiple of the line/entry height.

There are also some other changes and fixes to some of the modules. The layout of duplicates has changed; I have moved the add duplicate/virgin buttons to the bottom to match better with the layout of the history and snapshots modules. Some theming had to be adjusted to make sure items/lines are of fixed height (especially multiline labels as in metadata). @nilvus please comment on the changes.

If the changes here are liked, I should be able to extend the support to also cover GtkTextView and refactor all the other modules as well.

There is a remaining issue with updating the widgets after a resize. This is most obvious with the metadata module if there are some images with multiline metadata tags. While mouse moving over them, the size of the module will always be one step behind. I've tried adding all combinations of gtk_widget_queue_resize/draw/allocate to _scroll_wrap_resize, but so far haven't found a solution. I'd be very happy for someone to point out to me what I'm overlooking...

### CODE REVIEW

File: src/gui/gtk.c [hunk 1]
```diff
@@ -3066,6 +3066,110 @@ GtkWidget *dt_ui_notebook_page(GtkNotebook *notebook, const char *text, const ch
...
+
+  gint increment = _get_container_row_heigth(w);
+
+  gint height = dt_conf_get_int(config_str);
+
+  gint max_height = DT_PIXEL_APPLY_DPI(1000);
+
+  height = (height < 1) ? 1 : (height > max_height) ? max_height : height;
+
+  dt_conf_set_int(config_str, height);
```
Comment (johnny-bit) [inline src/gui/gtk.c:3109]: After my recent fun with dt conf you might replace this whole bit with `dt_conf_get_and_sanitize_int()`

**Review by johnny-bit — COMMENTED** [review]
I haven't tested nor did anything other than to suggest to use my recently introduced stuff :)

File: src/gui/gtk.c (hunk 1 above)
Comment (dterrahe) [reply src/gui/gtk.c:3109]: We are almost 100% in sync!

**Comment by Nilvus** [comment]
@dterrahe: a good direction there and I love seeing ctrl-scroll extended.

About duplicates modules, 3 things: 
- first, I think it's better with that PR, especially having label text taking all size and so delete icon on top right.
- but I find strange now to have number centered alone that way. I would suggest to put it on top left, on same line as delete cross
- some margins missing: on the right of thumbnails and on top of new buttons

For last one, here my CSS proposal to fix it (to integrate around line 580 in darktable.css, between buttons and entry settings):

```
#left #lib-plugin-ui .duplicate-ui entry 
{
  margin-left: 0.5em;
}

#left .duplicate-ui .text-button
{
  margin-top: 1em;
}
```

Feel free of course to integrate it on your PR.

**Comment by Nilvus** [comment]
Last thing: about 2nd issue, if you put number on top left, it should of course be vertically aligned to label left, so have same left margin as entry proposed on my CSS (if ok for you to integrate it).

**Comment by dterrahe** [comment]
Thanks for your review @Nilvus 
I put the number on the same line as the delete button as requested, but I find it slightly more pleasing looking centered; it is the same layout as top level windows, so the close button on the right also feels very natural this way. Have a look at the screen shot below and tell me what you think.
image

I've also reduced the 1em top margin for the buttons you suggested to 4px; now it is more consistent with the history and snapshot buttons (which are 2px margin and 2px padding).

Lastly, I've pushed a change to make scrolling with "mouse wheel scrolls modules" enabled behave the same as with the other scrolled-windows modules; if there is no scroll bar, the whole panel is scrolled. In the previous iteration of this PR, the mouse scroll would always be captured and not move on.
Ideally, in cases where there is a scrollbar visible, I would like the scrolling to be captured within the module until the end of the viewport is reached, and then I would like the panel to continue scrolling. But I don't know how to achieve this effect as it isn't what the scrollbars do by default. So I'll leave this behavior as before and elsewhere.

File: src/gui/gtk.c (hunk 1 above)
Comment (dterrahe) [reply src/gui/gtk.c:3109]: > After my recent fun with dt conf you might replace this whole bit with `dt_conf_get_and_sanitize_int()`

After the recent conf changes, what do you suggest I do here? Add 
`    <type min="1" max="1000">int</type>
`
to each of the entries in the xml and just do a simple dt_conf_get_int here?

**Comment by Nilvus** [comment]
> Thanks for your review @Nilvus
> I put the number on the same line as the delete button as requested, but I find it slightly more pleasing looking centered; it is the same layout as top level windows, so the close button on the right also feels very natural this way. Have a look at the screen shot below and tell me what you think.

Great for me that way. Centered on same line is much better than alone.

> I've also reduced the 1em top margin for the buttons you suggested to 4px; now it is more consistent with the history and snapshot buttons (which are 2px margin and 2px padding).

Didn't check that. I found that less aerate but indeed more consistent with other modules. So good for me. Thanks for your work.

File: src/gui/gtk.c (hunk 1 above)
Comment (johnny-bit) [reply src/gui/gtk.c:3109]: Later this weekend I'll propose a PR that will allow just that :)

Comment (johnny-bit) [reply src/gui/gtk.c:3109]: My PR is done now :) also I'd recommend more strict "min". `1` seems a little low for module window height :)

Comment (dterrahe) [reply src/gui/gtk.c:3109]: It gets rounded up to full row/element heights (depending on font etc). So 1 is intentional; there will always be one full row. I don't see a need to force it to be higher than that, if someone want a 1 row scroll window, why stop them?

**Comment by TurboGit** [comment]
When I open the style module in lighttable (closed when starting) I get this:

Capture d’écran de 2020-10-05 08-26-24

This gets properly redrawn as soon as I fly over the module with the mouse. So probably a missing redraw signal.

**Comment by TurboGit** [comment]
Same issue with history & mask module when entering darkrrom:

image

**Review by TurboGit — CHANGES_REQUESTED** [review]
Many modules does not get properly displayed until some refresh signal is received.

**Comment by dterrahe** [comment]
> Many modules does not get properly displayed until some refresh signal is received.

Thanks @TurboGit for looking at this.

I mentioned this problem in the submission comment to this PR. I have tried all kinds of gtk_widget_queue_resize/draw etc, on the widget, the scrolledwindow and parent, and even tried to force an immediate recalc/redraw with gtk_widget_size_allocate etc but nothing makes the window update itself. It is not a problem with the underlying box, grid or view; that gives the correct new size, so the size_request is putting in the required size. It is just the scrolledwindow not redrawing; the next redraw "fixes" it, but always a step behind. I'm even wondering if updates get temporarily disabled or postponed, because usually a size_request triggers an immediate repaint. If anybody has any other suggestions of what could fix this I would be very grateful. Who knows most about this? @AlicVB ?

**Comment by dterrahe** [comment]
After having tried everything, I did one more experiment, which I hadn't attempted before because it seemed obviously incorrect; the resize routine for the scrolled window now connects to the "draw" signal for the wrapped widget, rather than "size-allocate". Miraculously, now it all seems to work as intended.

@TurboGit Please have another look and let me know if this works for you and if you have any reservations. The draw signal must get called many more times than the size-allocate signal, but to my honest surprise I don't see any malfunctions.

**Comment by TurboGit** [comment]
> I mentioned this problem in the submission comment to this PR.

Right and I read it when the PR landed here and forgot :( Anyway good to see that you've found a solution to this problem. I'll have  a look.

**Review by TurboGit — APPROVED** [review]
Works for me. The "draw" signal is ok by me but I'm not expert on Gtk+ trick :) A nice addition to dt. Thanks.

**Pull request merged**